### PR TITLE
Development sshclient

### PR DESCRIPTION
**Pull request by RafyAmgadBenjamin**
working on issue https://github.com/threefoldtech/js-ng/issues/344
fixing issues in 
- sshkey
- sshclient
- jsync 
- syncer

### CODE REVIEW

File: jumpscale/tools/syncer/__init__.py [hunk 1]
```diff
@@ -198,12 +206,27 @@ def on_deleted(self, event):
...
+        j.logger.debug("Modified {}: {}".format(what, event.src_path))
 
-        # dest_path = self._rewrite_path_for_dest(event.src_path)
-        # j.logger.debug("will modify in {}".format(dest_path))
+        dest_path = self._rewrite_path_for_dest(event.src_path)
+        j.logger.debug("will modify in {}".format(dest_path))
+
+        for cl in self._get_sshclients():
+            if what == "directory":
+                j.logger.debug("A new file has been created")
```
Comment (xmonader) [inline jumpscale/tools/syncer/__init__.py:227]: can we add the filename here?

File: jumpscale/clients/sshclient/sshclient.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 """
 
-SSHClient modules helps connecting to a remote machine and 
+SSHClient modules helps connecting to a remote machine and
```
Comment (alichaddad) [inline jumpscale/clients/sshclient/sshclient.py:3]: Complete/fix the docstring.

File: jumpscale/tools/syncer/__init__.py [hunk 3]
```diff
@@ -166,13 +167,16 @@ def on_moved(self, event):
...
-            if j.sals.fs.is_file(dest_path):
-                cl.sftp.mkdir(j.sals.fs.parent(dest_path))
-                cl.sftp.put(event.dest_path, dest_path)
+            if not event.is_directory:
+                try:
+                    # in case file is moved
+                    cl.sshclient.sftp.put(event.dest_path, dest_path)
+                    cl.sshclient.run("rm {}".format(self._rewrite_path_for_dest(event.src_path)))
+                except:
+                    j.logger.debug(f"File {dest_path} was not able to be moved")
```
Comment (alichaddad) [inline jumpscale/tools/syncer/__init__.py:176]: Change it to "Ignoring {dest_path}. Path was not found"

File: jumpscale/tools/syncer/__init__.py [hunk 4]
```diff
@@ -184,10 +188,14 @@ def on_created(self, event):
...
+                cl.sshclient.run("mkdir -p {}".format(dest_path))
+                self.observer.schedule(self, event.src_path)
             else:
-                cl.sftp.mkdir(j.sals.fs.parent(dest_path))
-                cl.run("touch {}".format(dest_path))
+                try:
+                    cl.sshclient.run("mkdir -p {}".format(j.sals.fs.parent(dest_path)))
+                    cl.sshclient.run("touch {}".format(dest_path))
+                except:
+                    j.logger.debug(f"File {dest_path} was not able to be created")
```
Comment (alichaddad) [inline jumpscale/tools/syncer/__init__.py:198]: Same as above

File: jumpscale/tools/syncer/__init__.py [hunk 5]
```diff
@@ -198,12 +206,27 @@ def on_deleted(self, event):
...
         j.logger.debug("will delete in {}".format(dest_path))
         for cl in self._get_sshclients():
-            cl.run("rm {}".format(dest_path))
+            if what == "directory":
+                cl.sshclient.run("rm -rf {}".format(dest_path))
+            else:
+                try:
+                    cl.sshclient.run("rm {}".format(dest_path))
+                except:
+                    j.logger.debug(f"File {dest_path} was not able to be deleted")
```
Comment (alichaddad) [inline jumpscale/tools/syncer/__init__.py:215]: Same.

**Pull request merged**